### PR TITLE
Add custom field endpoints

### PR DESCRIPTION
**Pull request by jenssegers**
Adding custom field info and list endpoints. Initial version without events/todos/activities context. Also no "dynamic" multi select type.

### CODE REVIEW

**Review by mrtus — DISMISSED** [review]

**Review by WouterSioen — DISMISSED** [review]

**Review by JordyDeGroeve — APPROVED** [review]

File: src/01-general/custom-fields.apib [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+### customFieldDefinitions.list [GET /customFieldDefinitions.list]
+
+Get a list of all the definitions of custom fields.
+
++ Request (application/json;charset=utf-8)
+
+    + Attributes (object)
+        + filter (object, optional)
+            + ids: `74855f4a-2b61-429c-81d8-c79ad3675a76`,`1a68ac98-b41f-4c60-af12-895c43771385` (array[string], optional)
+            + context: `contact` (enum[string], optional) - The context of the custom field definition
```
Comment (marijnvandevoorde) [inline src/01-general/custom-fields.apib]: Not sure about name "context", but don't really have a better idea either.

full list:
'contact','company','form','event','product','offertex_lead_category_link','project','sale','invoice','meeting_report','subscription','werkbonnen','inv_invoice_item','ticket','milestone','project_thirdparty','meeting','callback','todo','pro_external_cost'

File: src/01-general/custom-fields.apib [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                    + url - URL
+                    + company - Company
+                    + contact - Contact
+                    + product - Product
+                    + user - User
+            + label (string)
+            + group (string)
+            + required: false (boolean)
+            + configuration (object)
+                + One Of
```
Comment (marijnvandevoorde) [inline src/01-general/custom-fields.apib:115]: Can also be nothing, right?

 :ok: by @WouterSioen on 20/04/2018 11:38:34

**Review by marijnvandevoorde — DISMISSED** [review]
List of 'context' is incomplete

File: src/01-general/custom-fields.apib [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                        + contact - Contact
+                        + product - Product
+                        + user - User
+                + label (string)
+                + group (string)
+                + required: false (boolean)
+                + configuration (object)
+                    + One Of
+                        + options: foo, bar (array[string])
+                        + start_value: 1 (number)
```
Comment (JordyDeGroeve) [inline src/01-general/custom-fields.apib]: B: The value is saved as the next auto increment value. Better call this: `next_value`
@WouterSioen @mrnnn 

 :ok: by @WouterSioen on 29/03/2018 08:42:34

File: src/01-general/custom-fields.apib [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                    + contact - Contact
+                    + product - Product
+                    + user - User
+            + label (string)
+            + group (string)
+            + required: false (boolean)
+            + configuration (object)
+                + One Of
+                    + options: foo, bar (array[string])
+                    + start_value: 1 (number)
```
Comment (JordyDeGroeve) [inline src/01-general/custom-fields.apib]: B: See remark in `customFieldDefinitions.list`

 :ok: by @WouterSioen on 29/03/2018 08:42:35

**Review by JordyDeGroeve — CHANGES_REQUESTED** [review]
Rename `start_value` to `next_value`

**Comment by WouterSioen** [comment]
Should it still be `one of` if there's only one possibility?

File: src/01-general/custom-fields.apib (hunk 1 above)
Comment (stivni) [reply src/01-general/custom-fields.apib]: `entity`, `object`, `on`, `owner`, ... :)

Comment (stivni) [reply src/01-general/custom-fields.apib]: let's only focus on what we want to release right now ...

 :ok: by @WouterSioen on 20/04/2018 11:38:32

**Comment by WouterSioen** [comment]
@JordyDeGroeve can this be unblocked? It's fully implemented and can thus be released imo

**Review by JordyDeGroeve — APPROVED** [review]

**Review by WouterSioen — DISMISSED** [review]

**Comment by WouterSioen** [comment]
LGTM

**Review by mrtus — DISMISSED** [review]

**Review by JordyDeGroeve — CHANGES_REQUESTED** [review]
`Filter` and `Sort` on **.list** are not available yet. What do we do? 
Wait for it to be implemented or removed this from the docs, merge and add it back later on?

@jenssegers, @stivni, @mrnnn, @WouterSioen

**Comment by WouterSioen** [comment]
PP: remove it from this PR, then merge this, and implement filtering and sorting later. This is already good enough to become documented endpoints, especially since companies.add already supports custom fields and requires required custom fields.

**Comment by WouterSioen** [comment]
@JordyDeGroeve sorting is btw implemented as in this documentation. Always ascending on label. Only `filters` are not implemented yet.

**Review by JordyDeGroeve — DISMISSED** [review]

**Review by WouterSioen — DISMISSED** [review]
LGTM

**Review by WouterSioen — DISMISSED** [review]

**Review by mrtus — DISMISSED** [review]

File: src/01-general/custom-fields.apib [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                + label (string)
+                + group (string)
+                + required: false (boolean)
+                + configuration (object)
+                    + One Of
+                        + options: foo, bar (array[string])
+
+### customFieldDefinitions.info [GET /customFieldDefinitions.info]
+
+Get a info about a specific custom field definition.
```
Comment (dietrichm) [inline src/01-general/custom-fields.apib]: PP: "Get info about..."

 :ok: by @WouterSioen on 08/05/2018 08:21:15

**Review by WouterSioen — APPROVED** [review]

**Review by JordyDeGroeve — APPROVED** [review]

**Review by dietrichm — APPROVED** [review]

**Pull request merged**